rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport.h

```diff
@@ -2,28 +2,245 @@
+}
+
+static inline void*
+align_(size_t __align, size_t __size, void*& __ptr, size_t& __space) noexcept
+{
+    const auto __intptr = reinterpret_cast<uintptr_t>(__ptr);
+    const auto __aligned = (__intptr - 1u + __align) & -__align;
+    const auto __diff = __aligned - __intptr;
+    if ((__size + __diff) > __space)
+        return nullptr;
+    else
+    {
+        __space -= __diff;
+        return __ptr = reinterpret_cast<void*>(__aligned);
+    }
+}
+
+template <typename MembersType>
+static size_t calculateMaxAlign(const MembersType *members)
+{
+    size_t max_align = 0;
+
+    for(unsigned long i = 0; i < members->member_count_; ++i)
+    {
+        size_t alignment = 0;
+        const auto &member = members->members_[i];
+
+        if(member.is_array_ && (!member.array_size_ || member.is_upper_bound_))
+        {
+            alignment = alignof(std::vector<unsigned char>);
+        }
+        else
+        {
+            switch(member.type_id_)
+            {
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BOOL:
+                    alignment = alignof(bool);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BYTE:
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT8:
+                    alignment = alignof(uint8_t);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_CHAR:
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT8:
+                    alignment = alignof(char);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_FLOAT32:
+                    alignment = alignof(float);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_FLOAT64:
+                    alignment = alignof(double);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT16:
+                    alignment = alignof(int16_t);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT16:
+                    alignment = alignof(uint16_t);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT32:
+                    alignment = alignof(int32_t);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT32:
+                    alignment = alignof(uint32_t);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT64:
+                    alignment = alignof(int64_t);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT64:
+                    alignment = alignof(uint64_t);
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING:
+                    alignment = alignof(std::string);
```

Comment:
I think it's not needed to specialize for `rosidl_generator_c__String` because all strings end up being serialized/deserialized to/from `std::string`. In the case of `rosidl_generator_c__String`, they are first converted to `std::string` before they are passed to `eprosima::fastcdr::Cdr`